### PR TITLE
Remove NoopGatewayAllocator in favor of a more realistic mock

### PR DESCRIPTION
**Pull request by bleskes**
Many of our unit tests instantiate an `AllocationService`, which requires having a `GatewayAllocator`. Today almost all of our test use a class called `NoopGatewayAllocator` which does nothing, effectively leaving all shard assignments to the balanced allocator. This is sad as it means we test a system that behaves differently than our production logic in very basic things. For example, a started primary that is lost will be assigned to a node that didn't use to have it.

This PR removes `NoopGatewayAllocator` in favor of a new `TestGatewayAllocator` that inherits the standard `GatewayAllocator` and overrides shard information fetching to return information based on historical assignments the allocator has done. The only exception is `BalanceConfigurationTests` which does test only the balancer and I opted to not have it work around the `GatewayAllocator` being in it's way.

The PR also fixes all resulting failures _plus_ one production failure for which I will open a dedicated PR (note - that fix is included here as well to make sure it works).

### CODE REVIEW

**Comment by bleskes** [comment]
@ywelsch I've merged master into this so #20638 will be remove from it.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A gateway allocator implementation that keeps an in memory list of started shard allocation
+ * that are used as replies to the, normally async, fetch data requests.
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: can you add a few sentences how it behaves under nodes that disappear / reappear.

**Review by ywelsch — COMMENTED** [review]
Left minor comments / questions. It will be interesting t o combine this with `IndicesClusterStateServiceRandomUpdatesTests` by introducing random failures into the TestGatewayAllocator to see how the system reacts to random failures during the shard fetching ;-)

File: core/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 2]
```diff
@@ -290,7 +290,7 @@ public String prettyPrint() {
             for (int shard = 0; shard < indexMetaData.getNumberOfShards(); shard++) {
                 sb.append(TAB).append(TAB).append(shard).append(": ");
                 sb.append("p_term [").append(indexMetaData.primaryTerm(shard)).append("], ");
-                sb.append("a_ids ").append(indexMetaData.inSyncAllocationIds(shard)).append("\n");
+                sb.append("sa_ids ").append(indexMetaData.inSyncAllocationIds(shard)).append("\n");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/ClusterState.java]: synced allocations, but I agree that `isa_ids` are better.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 3]
```diff
@@ -138,7 +138,8 @@ public UnassignedShardDecision makeAllocationDecision(final ShardRouting unassig
 
         if (inSyncAllocationIds.isEmpty()) {
             assert Version.indexCreated(indexMetaData.getSettings()).before(Version.V_5_0_0_alpha1) :
-                "trying to allocated a primary with an empty allocation id set, but index is new";
+                "trying to allocated a primary with an empty in sync allocation id set, but index is new. index: "
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: will changed.

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/FilterAllocationDeciderTests.java [hunk 4]
```diff
@@ -86,57 +90,80 @@ public void testFilterInitialRecovery() {
...
         assertEquals(filterAllocationDecider.canAllocate(routingTable.index("idx").shard(0).shards().get(0),
             state.getRoutingNodes().node("node1")
-            ,allocation), Decision.YES);
+            , allocation), Decision.YES);
     }
 
     private ClusterState createInitialClusterState(AllocationService service, Settings settings) {
-        boolean shrinkIndex = randomBoolean();
+        RecoverySource.Type recoveryType = randomFrom(RecoverySource.Type.EMPTY_STORE,
+            RecoverySource.Type.LOCAL_SHARDS, RecoverySource.Type.SNAPSHOT);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/FilterAllocationDeciderTests.java]: I opted to do it like it is now for two reasons - it's more explicit so people reading can know what's going on and it offers an external validation of what's in INITIAL_RECOVERY_TYPES.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 5]
```diff
@@ -78,11 +81,10 @@ public void buildTestAllocator() {
...
         final RoutingAllocation allocation;
-        if (randomBoolean()) {
-            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), randomBoolean(), Version.CURRENT);
-        } else {
-            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), true, Version.V_2_1_0);
-        }
+        // with old version, we can't know if a shard was allocated before or not
+        allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(),
+            INDEX_REOPENED, Version.CURRENT);
+        //randomFrom(INDEX_CREATED, CLUSTER_RECOVERED, INDEX_REOPENED)
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: good catch. left over from debugging. I put it back into it's place.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 6]
```diff
@@ -616,17 +624,27 @@ public void testEnoughCopiesFoundForAllocationOnLegacyIndexWithDifferentVersion(
...
+                routingTableBuilder.addAsNew(metaData.index(shardId.getIndex()));
+                break;
+            case CLUSTER_RECOVERED:
+                routingTableBuilder.addAsRecovery(metaData.index(shardId.getIndex()));
+                break;
+            case INDEX_REOPENED:
+                routingTableBuilder.addAsFromCloseToOpen(metaData.index(shardId.getIndex()));
+                break;
+            default:
+                throw new UnsupportedOperationException("can't do " + reason + " for you. teach me");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Now, this is an interesting philosophical point - does one take ownership of it's shortcoming or is the fault of other people :) . On a more serious note - in this case I think the other paths should be implemented and a valid thing to expect here. Hence the choice of exception.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java (hunk 1 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: sure.

**Comment by abeyad** [comment]
@bleskes in #20626 and #20642, `StartedRerouteAllocation` and `FailedRerouteAllocation` were removed, the consequence to your PR being that `GatewayAllocator#applyStartedShards` now takes `(RoutingAllocation, List<ShardRouting> startedShards)` and `GatewayAllocator#applyFailedShards` now takes `(RoutingAllocation, List<FailedShard> failedShards)`

**Comment by bleskes** [comment]
@abeyad yep saw it. Thanks for the extra ping.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * For now only primary shard related data is fetched. Replica request always get an empty response.
+ *
+ * This class is useful to use in unit tests that require the functionality of {@link GatewayAllocator} but do
+ * not have all the infrastructure required to use it.
+ */
+public class TestGatewayAllocator extends GatewayAllocator {
+
+    Map<String, Map<ShardId,ShardRouting>> knownAllocations = new HashMap<>();
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: you have been spending too much time with the TLA model :) . Of course, you are right but I opted to go for clarity and simplicity here. I think it's the right thing to optimize for here.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    Map<String, DiscoveryNode> knownNodes = new HashMap<>();
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: that's a great idea.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    PrimaryShardAllocator primaryShardAllocator = new PrimaryShardAllocator(Settings.EMPTY) {
+        @Override
+        protected AsyncShardFetch.FetchResult<NodeGatewayStartedShards>
+        fetchData(ShardRouting shard, RoutingAllocation allocation) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: yeah. fixed.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        @Override
+        protected AsyncShardFetch.FetchResult<NodeStoreFilesMetaData> fetchData(ShardRouting shard, RoutingAllocation allocation) {
+            // for now, just pretend no node has data
+            final ShardId shardId = shard.shardId();
+            return new AsyncShardFetch.FetchResult<>(shardId, Collections.emptyMap(), Collections.emptySet(),
+                allocation.getIgnoreNodes(shardId));
+        }
+    };
+
+    public TestGatewayAllocator() {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: it is the first method. the rest are fields. I personally don't care. let me know what you prefer.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+    @Override
+    public void allocateUnassigned(RoutingAllocation allocation) {
+        updateNodes(allocation);
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        unassigned.sort(PriorityComparator.getAllocationComparator(allocation)); // sort for priority ordering
+
+        primaryShardAllocator.allocateUnassigned(allocation);
+        replicaShardAllocator.processExistingRecoveries(allocation);
+        replicaShardAllocator.allocateUnassigned(allocation);
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: +1. done.

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java (hunk 10 above)
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: right, I missed that. Ignore my comment.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 6 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Here are some simple rules I try to follow when selecting the right exception type:

http://smallwig.blogspot.com/2008/06/common-java-unchecked-exception-types.html

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java (hunk 11 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: I now remember why I did it this way - I couldn't find a good way to get all the inner classes to work as they can not be constructed and passed around as a parameter. I took another approach with https://github.com/elastic/elasticsearch/pull/20637/commits/13b5c5c3272aedbc631d578d1e585886bef817da

**Comment by bleskes** [comment]
@ywelsch thanks for the review. I merged from master (to get all the changes in) and pushed some more commits as a fall out of the changes you suggested. Can you take another look?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 6 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: I agree with @ywelsch here; I had not seen the advice that he linked to but it's basically the same as in Joshua Bloch's Effective Java (except he takes a stand on NPE vs. IAE: throw NPE).

File: test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java (hunk 11 above)
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java]: ah, that's unfortunate :-/

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 6 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: what ever makes you guys happy :)

**Review by ywelsch — APPROVED** [review]
LGTM, thanks @bleskes!

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 6 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Better commit messages would make some people happy. :wink:

**Comment by bleskes** [comment]
@ywelsch I'm holding on pushing this. The extra assertion in IndexRoutingTable#validate has surfaced another issue with shadow replicas that I want to talk about (around not reusing aid with reinitlizing a replica as a primary - see [1]). I'm leaning towards removing this assertion for now so we can get this in and follow up after discussion.

[1] look for the only caller of `RoutingChangesObserver#startedPrimaryReinitialized`

**Comment by ywelsch** [comment]
I'm good if you add `&& IndexMetaData.isIndexUsingShadowReplicas(indexMetaData.getSettings()) == false` to the assertion, get this in and open an issue for the shadow replica case.

**Pull request merged**